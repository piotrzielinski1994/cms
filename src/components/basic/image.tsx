import { cn } from '@/utils/tailwind';
import NextImage, { ImageProps as NextImageProps } from 'next/image';
import defaultTheme from 'tailwindcss/defaultTheme';

type ImageProps = {
  src: string;
  alt: string;
  className?: string;
  aspectRatio?: {
    width: number;
    height: number;
  };
};

const sizesPerBreakpoint = Object.values(defaultTheme.screens)
  .map((it) => parseInt(it))
  .sort((a, b) => a - b)
  .map((it) => `(max-width: ${it}px) 100vw`)
  .join(', ');

const Image = ({ aspectRatio, className, ...props }: ImageProps) => {
  const sizing: Partial<NextImageProps> = aspectRatio ?? {
    width: 0,
    height: 0,
    className: 'w-full h-auto',
  };

  return (
    <NextImage
      objectFit="cover"
      sizes={sizesPerBreakpoint}
      {...props}
      {...sizing}
      className={cn(sizing.className, 'cms-image', className)}
    />
  );
};

export { Image };
