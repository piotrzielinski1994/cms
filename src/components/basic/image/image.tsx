import { cssVariables } from '@/_old/cssVariables';
import NextImage from 'next/image';

const { breakpoints } = cssVariables;

const Image = (props: any) => {
  const sizes = false
    ? props.sizes
    : Object.entries(breakpoints)
        .map(([, value]) => `(max-width: ${value}px) ${value * 2}w`)
        .join(', ');

  return (
    <picture className="relative">
      <NextImage
        src={props.url}
        alt={props.alt}
        fill={!props.width}
        width={props.width}
        height={props.height}
        // sizes="100vw"
        sizes={sizes}
        placeholder="blur"
        blurDataURL="/images/shapes-1-100x42.jpg"
      />
    </picture>
  );
};

export { Image };
