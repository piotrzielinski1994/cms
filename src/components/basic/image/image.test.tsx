import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { axe } from 'vitest-axe';
import { Image } from './image';

describe('Image', () => {
  it('should have no accessibility violations', async () => {
    const { container } = render(<Image src="/image.webp" alt="Label" />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const { container } = render(<Image src="/image.webp" alt="Label" />);
    expect(container).toMatchSnapshot();
  });
});
