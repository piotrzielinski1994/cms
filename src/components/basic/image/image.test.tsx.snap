// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Image > should match the snapshot 1`] = `
<div>
  <img
    alt=""
    class="w-full h-auto object-cover cms-image"
    data-nimg="1"
    decoding="async"
    height="0"
    loading="lazy"
    sizes="100vw"
    src="/_next/image?url=%2Fimage.webp&w=3840&q=75"
    srcset="/_next/image?url=%2Fimage.webp&w=640&q=75 640w, /_next/image?url=%2Fimage.webp&w=750&q=75 750w, /_next/image?url=%2Fimage.webp&w=828&q=75 828w, /_next/image?url=%2Fimage.webp&w=1080&q=75 1080w, /_next/image?url=%2Fimage.webp&w=1200&q=75 1200w, /_next/image?url=%2Fimage.webp&w=1920&q=75 1920w, /_next/image?url=%2Fimage.webp&w=2048&q=75 2048w, /_next/image?url=%2Fimage.webp&w=3840&q=75 3840w"
    style="color: transparent;"
    width="0"
  />
</div>
`;
